File: site/python/15.py
grid = 20
routes = [[0]*(grid+1) for i in range(grid+1)]

def branchSum(x,y):
    if x>grid or y> grid:
        return 0
    if x == grid or y == grid:
        r = x if y ==  grid else y
        if r+1 == grid:
            return 1

    if routes[x][y] > 0:
        return routes[x][y]

    routes[x][y]=branchSum(x+1,y)+branchSum(x,y+1)

    return routes[x][y]

print(branchSum(0,0))

File: site/javascript/15.js
const grid = 20;
let routes = new Array(grid+1).fill(0)
routes = routes.map(x=> x = new Array(grid+1).fill(0))

function branchSum(x,y){
    if(x>grid || y>grid){
        return 0
    }
    if(x == grid || y == grid){
        let r = x == grid ? y : x;
        if(r+1 == grid ){
            return 1
        }        
    }
    if(routes[x][y]!=0) return routes[x][y]
    return routes[x][y]=branchSum(x+1,y) + branchSum(x,y+1)

}


console.log(branchSum(0,0));
console.table(routes)
